File: Homebrew-OA/cloudPublisher/src/my_pcl_tutorial/src/FileReader.cpp
#include <ros/console.h>
#include <ros/ros.h>
#include <std_msgs/String.h>
#include <pcl_conversions/pcl_conversions.h>
#include <pcl/point_cloud.h>
#include <pcl/point_types.h>
#include <sensor_msgs/PointCloud2.h>
#include <pcl/io/pcd_io.h>

using namespace std;
using namespace ros;

int main(int argc, char **argv){
	// Initialzie ROS, allows remapping through command line
	ros::init(argc, argv, "chatter");

	ros::NodeHandle n;

	// Returns a ros Publisher object
	ros::Publisher lidarScanPub = n.advertise<sensor_msgs::PointCloud2>("chatter", 10);

	ros::Rate loopRate(1);

	while(ros::ok()){

		// Create pointcloud object
		sensor_msgs::PointCloud2 cloud_blob;

		// Load the file
		if (pcl::io::loadPCDFile("/home/ethan/Projects/Obstacle_Avoidance_2024-2025/Homebrew-OA/cloudPublisher/src/my_pcl_tutorial/src/sampleData/office1.pcd", cloud_blob) == -1) {
			PCL_ERROR ("Couldn't read file\n");
			return(-1);
		}

		cloud_blob.header.frame_id = "map";

		// Publish point cloud
		lidarScanPub.publish(cloud_blob);

		loopRate.sleep();

	}
	return 1;
}

File: Homebrew-OA/catkin_ws/src/pc_package/scripts/FileReader.py
#!/usr/bin/env python3

import rospy
from std_msgs.msg import String
import sensor_msgs.point_cloud2 as pc2
from sensor_msgs.msg import PointCloud2, PointField
import open3d as o3d
import numpy as np
import math
def main():
    # Initialize the ROS node
    rospy.init_node('chatter', anonymous=True)

    # Publisher for the PointCloud2 topic
    lidar_scan_pub = rospy.Publisher('pointcloud', PointCloud2, queue_size=10)

    # Set loop rate to 10Hz
    rate = rospy.Rate(10)
    
    cloud_o3d = o3d.io.read_point_cloud("/home/ethan/Projects/Obstacle_Avoidance_2024-2025/Homebrew-OA/cloudPublisher/src/my_pcl_tutorial/src/sampleData/office1.pcd")
    points = np.asarray(cloud_o3d.points)
    # Create a list of tuples in (x, y, z) format
    ros_points = [(p[0], p[1], p[2]) for p in points]

    cloud_msg = pc2.create_cloud_xyz32(rospy.Header(frame_id="map"), ros_points)

    while not rospy.is_shutdown():

        # Publish the point cloud
        lidar_scan_pub.publish(cloud_msg)
        
        # Sleep to maintain loop rate
        rate.sleep()
    
    return

if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass